base/gmp.jl

```diff
@@ -308,21 +308,21 @@ BigInt(x::Float16) = BigInt(Float64(x))
 BigInt(x::Float32) = BigInt(Float64(x))
 
 function BigInt(x::Integer)
-    x == 0 && return BigInt(Culong(0))
+    # On 64-bit Windows, `Clong` is `Int32`, not `Int64`, so construction of
+    # `Int64` constants, e.g. `BigInt(3)`, uses this method.
+    typemin(Clong) <= x <= typemax(Clong) && return BigInt(x % Clong)
```

Comment:
Maybe check `isbits(x)` here? I could imagine for custom bigint types, these comparisons would be more expensive than the conversion.